From 2bb904b7f49d86952b4c66863c3860f7cda65e61 Mon Sep 17 00:00:00 2001
From: James Lee <jlee@thestaticvoid.com>
Date: Fri, 11 Mar 2011 23:01:02 -0500
Subject: [PATCH] Fixes for Sun Studio compilation.

This commit makes several changes to allow the program to compile under
Sun Studio (on Solaris).  Sun Studio is the de facto C++ compiler on
Solaris and presents several challenges compared to GCC.  For example,
it has different command line switches, more strict support for the
standards (including fewer language extensions), and incomplete support
for the C++ standard library using the libCrun and libCstd libraries.

Here is a summary of the changes:

src/CMakeLists.txt:
- Conditionally set command line flags based on compiler (GCC or Sun
  Studio)
- Link libdarktable.so as a C++ library (it contains C++ objects exif.o
  and imageio_exr.o whose exception handling fails to work when linked
  as a C library).
- Link darktable against libm as it contains calls to logf.

src/LibRaw/CMakeLists.txt:
src/rawspeed/CMakeLists.txt:
- Set PIC flag based on compiler (-fPIC for GCC, -Kpic for Sun Studio)

src/LibRaw/libraw/libraw_datastream.h:
- libCstd auto_ptr doesn't have assignment operator...see:
  http://developers.sun.com/solaris/articles/cmp_stlport_libCstd.html

src/common/darktable.c:
- Fix invalid call to fprintf.

src/common/dtpthread.h:
- Include glib.h for call to g_strlcpy.

src/control/jobs/control_jobs.c:
- Sun Studio doesn't like 'stderr' in OpenMP directives, so temporarily
  reassign it.

src/iop/atrous.c:
- Include emmintrin.h instead of xmmentrin.h, fixing implicit definition
  of _mm_set1_epi32.
- Add missing variable to OpenMP directive.

src/iop/bilateral.cc:
- Struct initialization is a C99ism that isn't supported under C++.

src/iop/demosaic.c:
- Add missing variable to OpenMP directive.
- __builtin_expect is a GCC extension.

src/iop/temperature.c:
- Add missing variable to OpenMP directive.

src/iop/tonemap.cc:
- Struct initialization is a C99ism that isn't supported under C++.

src/rawspeed/RawSpeed/*.h:
- Replace non-standard "#pragma once" with normal header guards.  This
  change needs to be pushed upstream.

tools/create_metadata.sh:
- Solaris 'echo' doesn't have '-n' flag.  Use 'printf' instead.
---
 src/CMakeLists.txt                              |   72 ++++++++++++++--------
 src/LibRaw/CMakeLists.txt                       |    9 +++-
 src/LibRaw/libraw/libraw_datastream.h           |    4 +-
 src/common/darktable.c                          |    4 +-
 src/common/dtpthread.h                          |    1 +
 src/control/jobs/control_jobs.c                 |    9 ++-
 src/iop/atrous.c                                |    4 +-
 src/iop/bilateral.cc                            |   14 +++--
 src/iop/demosaic.c                              |    6 ++-
 src/iop/temperature.c                           |    2 +-
 src/iop/tonemap.cc                              |    8 +-
 src/rawspeed/CMakeLists.txt                     |    9 +++-
 src/rawspeed/RawSpeed/ArwDecoder.h              |    6 ++-
 src/rawspeed/RawSpeed/BitPumpJPEG.h             |    6 ++-
 src/rawspeed/RawSpeed/BitPumpMSB.h              |    6 ++-
 src/rawspeed/RawSpeed/BitPumpMSB32.h            |    6 ++-
 src/rawspeed/RawSpeed/BitPumpPlain.h            |    6 ++-
 src/rawspeed/RawSpeed/BlackArea.h               |    6 ++-
 src/rawspeed/RawSpeed/ByteStream.h              |    6 ++-
 src/rawspeed/RawSpeed/ByteStreamSwap.h          |    6 ++-
 src/rawspeed/RawSpeed/Camera.h                  |    6 ++-
 src/rawspeed/RawSpeed/CameraMetaData.h          |    6 ++-
 src/rawspeed/RawSpeed/CameraMetadataException.h |    6 ++-
 src/rawspeed/RawSpeed/ColorFilterArray.h        |    6 ++-
 src/rawspeed/RawSpeed/Common.h                  |    6 ++-
 src/rawspeed/RawSpeed/Cr2Decoder.h              |    7 ++-
 src/rawspeed/RawSpeed/DngDecoder.h              |    6 ++-
 src/rawspeed/RawSpeed/DngDecoderSlices.h        |    6 ++-
 src/rawspeed/RawSpeed/FileIOException.h         |    6 ++-
 src/rawspeed/RawSpeed/FileMap.h                 |    6 ++-
 src/rawspeed/RawSpeed/FileReader.h              |    6 ++-
 src/rawspeed/RawSpeed/IOException.h             |    6 ++-
 src/rawspeed/RawSpeed/LJpegDecompressor.h       |    6 ++-
 src/rawspeed/RawSpeed/LJpegPlain.h              |    6 ++-
 src/rawspeed/RawSpeed/NefDecoder.h              |    7 ++-
 src/rawspeed/RawSpeed/NikonDecompressor.h       |    6 ++-
 src/rawspeed/RawSpeed/OrfDecoder.h              |    6 ++-
 src/rawspeed/RawSpeed/PefDecoder.h              |    6 ++-
 src/rawspeed/RawSpeed/PentaxDecompressor.h      |    6 ++-
 src/rawspeed/RawSpeed/RawDecoder.h              |    6 ++-
 src/rawspeed/RawSpeed/RawDecoderException.h     |    6 ++-
 src/rawspeed/RawSpeed/RawImage.h                |    6 ++-
 src/rawspeed/RawSpeed/Rw2Decoder.h              |    6 ++-
 src/rawspeed/RawSpeed/TiffEntry.h               |    6 ++-
 src/rawspeed/RawSpeed/TiffEntryBE.h             |    6 ++-
 src/rawspeed/RawSpeed/TiffIFD.h                 |    6 ++-
 src/rawspeed/RawSpeed/TiffIFDBE.h               |    6 ++-
 src/rawspeed/RawSpeed/TiffParser.h              |    6 ++-
 src/rawspeed/RawSpeed/TiffParserException.h     |    6 ++-
 src/rawspeed/RawSpeed/TiffParserHeaderless.h    |    6 ++-
 src/rawspeed/RawSpeed/TiffParserOlympus.h       |    6 ++-
 tools/create_metadata.sh                        |    8 +-
 52 files changed, 292 insertions(+), 94 deletions(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 7aa3d68..c1cd81e 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -65,7 +65,6 @@ include_directories(${CMAKE_CURRENT_SOURCE_DIR}/LibRaw)
 
 # initial compiler flags
 add_definitions("-DHAVE_CONFIG_H")
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c99")
 
 # Check for openmp
 find_package(OpenMP)
@@ -195,32 +194,53 @@ endif(GPHOTO2_FOUND)
 #
 # Set default compile flags for GCC
 #
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -fno-strict-aliasing")
-# c++ code doesn't like -Werror (libraw)
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -fno-strict-aliasing")
-
-if(NOT APPLE)
-  # apple build fails on deprecated warnings..
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Werror")
-endif(NOT APPLE)
+if(CMAKE_COMPILER_IS_GNUCC)
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=c99 -Wall -fno-strict-aliasing")
+  # c++ code doesn't like -Werror (libraw)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -fno-strict-aliasing")
+  
+  if(NOT APPLE)
+    # apple build fails on deprecated warnings..
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Werror")
+  endif(NOT APPLE)
+  
+  if(NOT WIN32)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -pthread")
+  endif(NOT WIN32)
+  
+  if(NOT BINARY_PACKAGE_BUILD)
+  	set(MARCH "-march=native")	
+  else()
+  	set(MARCH "-mtune=generic")
+  endif()
+  
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -msse2 -mfpmath=sse ${MARCH} -g")
+  set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O3 -ffast-math -fexpensive-optimizations")
+  set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -O0 -g -D_DEBUG")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -msse2 -mfpmath=sse ${MARCH} -g")
+  set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -ffast-math -fexpensive-optimizations")
+  set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -O0 -g -D_DEBUG")
+endif(CMAKE_COMPILER_IS_GNUCC)
 
-if(NOT WIN32)
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -pthread")
-endif(NOT WIN32)
 
-if(NOT BINARY_PACKAGE_BUILD)
-	set(MARCH "-march=native")	
-else()
-	set(MARCH "-mtune=generic")
-endif()
+#
+# Set default compile flags for Sun Studio 
+#
+if("${CMAKE_C_COMPILER_ID} ${CMAKE_CXX_COMPILER_ID}" MATCHES SunPro)
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -xc99 -features=extensions -xarch=sse2 -D__SSE2__ -D__unix__")
+  set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -fast")
+  set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -D_DEBUG")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -features=extensions -xarch=sse2 -D__SSE2__ -D__unix__")
+  set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -fast")
+  set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -D_DEBUG")
+
+  # -xopenmp normally forces -xO3...disable optimizations for debug
+  if(CMAKE_BUILD_TYPE MATCHES Debug)
+    string(REPLACE "-xopenmp" "-xopenmp=noopt" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}") 
+    string(REPLACE "-xopenmp" "-xopenmp=noopt" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}") 
+  endif(CMAKE_BUILD_TYPE MATCHES Debug)
+endif("${CMAKE_C_COMPILER_ID} ${CMAKE_CXX_COMPILER_ID}" MATCHES SunPro)
 
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -msse2 -mfpmath=sse ${MARCH} -g")
-set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O3 -ffast-math -fexpensive-optimizations")
-set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -O0 -g -D_DEBUG")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -msse2 -mfpmath=sse ${MARCH} -g")
-set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3 -ffast-math -fexpensive-optimizations")
-set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -O0 -g -D_DEBUG")
-  
 
 #
 # Generate config.h
@@ -259,7 +279,7 @@ if(APPLE)
   set_target_properties(lib_darktable PROPERTIES INSTALL_NAME_DIR "@executable_path/../lib/darktable")
 endif(APPLE)
 set_target_properties(lib_darktable PROPERTIES OUTPUT_NAME darktable)
-set_target_properties(lib_darktable PROPERTIES LINKER_LANGUAGE C)
+set_target_properties(lib_darktable PROPERTIES LINKER_LANGUAGE CXX)
 
 
 #
@@ -313,7 +333,7 @@ set_target_properties(darktable PROPERTIES CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)
 set_target_properties(darktable PROPERTIES CMAKE_INSTALL_RPATH_USE_LINK_PATH FALSE)
 set_target_properties(darktable PROPERTIES INSTALL_RPATH $ORIGIN/../${LIB_INSTALL}/darktable)
 set_target_properties(darktable PROPERTIES LINKER_LANGUAGE C)
-target_link_libraries(darktable ${LIBS})
+target_link_libraries(darktable ${LIBS} m)
 install(TARGETS darktable DESTINATION bin)
 
 
diff --git a/src/LibRaw/CMakeLists.txt b/src/LibRaw/CMakeLists.txt
index e5f7e18..1bf327f 100644
--- a/src/LibRaw/CMakeLists.txt
+++ b/src/LibRaw/CMakeLists.txt
@@ -5,7 +5,14 @@ set(LIBRAW_SOURCES internal/dcraw_common.cpp src/libraw_cxx.cpp internal/dcraw_f
 #
 # build liblibraw_r
 #
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
+if(CMAKE_COMPILER_IS_GNUCC)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
+endif(CMAKE_COMPILER_IS_GNUCC)
+
+
+if("${CMAKE_C_COMPILER_ID} ${CMAKE_CXX_COMPILER_ID}" MATCHES SunPro)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Kpic")
+endif("${CMAKE_C_COMPILER_ID} ${CMAKE_CXX_COMPILER_ID}" MATCHES SunPro)
 
 add_library(libraw_r STATIC ${LIBRAW_SOURCES})
 target_link_libraries(libraw_r)
diff --git a/src/LibRaw/libraw/libraw_datastream.h b/src/LibRaw/libraw/libraw_datastream.h
index 00b370e..2ca2efb 100644
--- a/src/LibRaw/libraw/libraw_datastream.h
+++ b/src/LibRaw/libraw/libraw_datastream.h
@@ -93,7 +93,7 @@ class LibRaw_file_datastream: public LibRaw_abstract_datastream
             std::auto_ptr<std::filebuf> buf(new std::filebuf());
             buf->open(filename, std::ios_base::in | std::ios_base::binary);
             if (buf->is_open()) {
-                f = buf;
+                f = std::auto_ptr<std::streambuf>(buf.release());
             }
         }
     }
@@ -172,7 +172,7 @@ class LibRaw_file_datastream: public LibRaw_abstract_datastream
             f = saved_f;
             return ENOENT;
         } else {
-            f = buf;
+            f = std::auto_ptr<std::streambuf>(buf.release());
         }
 
         return 0;
diff --git a/src/common/darktable.c b/src/common/darktable.c
index 2d46bc2..c987681 100644
--- a/src/common/darktable.c
+++ b/src/common/darktable.c
@@ -65,8 +65,8 @@ static int usage(const char *argv0)
 int dt_init(int argc, char *argv[])
 {
 #ifndef __SSE2__
-  fprintf("[dt_init] unfortunately we depend on SSE2 instructions at this time.\n");
-  fprintf("[dt_init] please contribute a backport patch (or buy a newer processor).\n");
+  fprintf(stderr, "[dt_init] unfortunately we depend on SSE2 instructions at this time.\n");
+  fprintf(stderr, "[dt_init] please contribute a backport patch (or buy a newer processor).\n");
   return 1;
 #endif
   bindtextdomain (GETTEXT_PACKAGE, DARKTABLE_LOCALEDIR);
diff --git a/src/common/dtpthread.h b/src/common/dtpthread.h
index aad66ae..a038e6e 100644
--- a/src/common/dtpthread.h
+++ b/src/common/dtpthread.h
@@ -24,6 +24,7 @@
 #include <stdlib.h>
 #include <stdio.h>
 #include <string.h>
+#include <glib.h>
 
 #ifdef _DEBUG
 
diff --git a/src/control/jobs/control_jobs.c b/src/control/jobs/control_jobs.c
index 9396b62..9f52862 100644
--- a/src/control/jobs/control_jobs.c
+++ b/src/control/jobs/control_jobs.c
@@ -448,14 +448,17 @@ int32_t dt_control_export_job_run(dt_job_t *job)
   dt_gui_background_jobs_can_cancel (j,job);
 
   double fraction=0;
+
+  // Sun Studio won't accpet 'stderr' in openmp directive, so redefine it
+  FILE *tmp_stderr = stderr;
 #ifdef _OPENMP
   // limit this to num threads = num full buffers - 1 (keep one for darkroom mode)
   // use min of user request and mipmap cache entries
   const int full_entries = dt_conf_get_int ("mipmap_cache_full_images");
   // GCC won't accept that this variable is used in a macro, considers
   // it set but not used, which makes for instance Fedora break.
-  const __attribute__((__unused__)) int num_threads = MAX(1, MIN(full_entries, darktable.mipmap_cache->num_entries[DT_IMAGE_FULL]) - 1);
-#pragma omp parallel default(none) private(imgid, size) shared(j, fraction, stderr, w, h, mformat, mstorage, t, sdata, job) num_threads(num_threads)
+  __attribute__((__unused__)) int num_threads = MAX(1, MIN(full_entries, darktable.mipmap_cache->num_entries[DT_IMAGE_FULL]) - 1);
+#pragma omp parallel default(none) private(imgid, size) shared(j, fraction, tmp_stderr, w, h, mformat, mstorage, t, sdata, job) num_threads(num_threads)
   {
 #endif
     // get a thread-safe fdata struct (one jpeg struct per thread etc):
@@ -484,7 +487,7 @@ int32_t dt_control_export_job_run(dt_job_t *job)
         if(!g_file_test(imgfilename, G_FILE_TEST_IS_REGULAR))
         {
           dt_control_log(_("image `%s' is currently unavailable"), image->filename);
-          fprintf(stderr, _("image `%s' is currently unavailable"), imgfilename);
+          fprintf(tmp_stderr, _("image `%s' is currently unavailable"), imgfilename);
           // dt_image_remove(imgid);
           dt_image_cache_release(image, 'r');
         }
diff --git a/src/iop/atrous.c b/src/iop/atrous.c
index 471400b..4a9226f 100644
--- a/src/iop/atrous.c
+++ b/src/iop/atrous.c
@@ -26,7 +26,7 @@
 #include "control/control.h"
 #include <memory.h>
 #include <stdlib.h>
-#include <xmmintrin.h>
+#include <emmintrin.h>
 // SSE4 actually not used yet.
 // #include <smmintrin.h>
 
@@ -134,7 +134,7 @@ eaw_decompose (float *const out, const float *const in, float *const detail, con
   const float filter[5] = {1.0f/16.0f, 4.0f/16.0f, 6.0f/16.0f, 4.0f/16.0f, 1.0f/16.0f};
 
 #ifdef _OPENMP
-#pragma omp parallel for default(none) schedule(static)
+#pragma omp parallel for default(none) schedule(static) shared(filter)
 #endif
   for(int j=0; j<height; j++)
   {
diff --git a/src/iop/bilateral.cc b/src/iop/bilateral.cc
index f4b5509..3089d93 100644
--- a/src/iop/bilateral.cc
+++ b/src/iop/bilateral.cc
@@ -269,12 +269,14 @@ int flags()
     module->priority = 270;
     module->params_size = sizeof(dt_iop_bilateral_params_t);
     module->gui_data = NULL;
-    dt_iop_bilateral_params_t tmp = (dt_iop_bilateral_params_t)
-    {
-      {
-        15.0, 15.0, 0.005, 0.005, 0.005
-      }
-    };
+
+    dt_iop_bilateral_params_t tmp;
+    tmp.sigma[0] = 15.0;
+    tmp.sigma[1] = 15.0;
+    tmp.sigma[2] = 0.005;
+    tmp.sigma[3] = 0.005;
+    tmp.sigma[4] = 0.005;
+
     memcpy(module->params, &tmp, sizeof(dt_iop_bilateral_params_t));
     memcpy(module->default_params, &tmp, sizeof(dt_iop_bilateral_params_t));
   }
diff --git a/src/iop/demosaic.c b/src/iop/demosaic.c
index 4b92263..197e986 100644
--- a/src/iop/demosaic.c
+++ b/src/iop/demosaic.c
@@ -175,7 +175,7 @@ pre_median_b(float *out, const float *const in, const dt_iop_roi_t *const roi, c
   for (int pass=0; pass < num_passes; pass++)
   {
 #ifdef _OPENMP
-#pragma omp parallel for default(none) shared(out) schedule(static)
+#pragma omp parallel for default(none) shared(out,lim) schedule(static)
 #endif
     for (int row=3; row<roi->height-3; row++)
     {
@@ -464,7 +464,11 @@ demosaic_ppg(float *out, const float *in, dt_iop_roi_t *roi_out, const dt_iop_ro
       float *color = (float *)&col;
       // fill all four pixels with correctly interpolated stuff: r/b for green1/2
       // b for r and r for b
+#ifdef __GNUC__
       if(__builtin_expect(c & 1, 1)) // c == 1 || c == 3)
+#else
+      if(c & 1)
+#endif
       {
         // calculate red and blue for green pixels:
         // need 4-nbhood:
diff --git a/src/iop/temperature.c b/src/iop/temperature.c
index 8610637..62a7dad 100644
--- a/src/iop/temperature.c
+++ b/src/iop/temperature.c
@@ -169,7 +169,7 @@ void process (struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, void
   {
     const float coeffsi[3] = {d->coeffs[0]/65535.0f, d->coeffs[1]/65535.0f, d->coeffs[2]/65535.0f};
 #ifdef _OPENMP
-#pragma omp parallel for default(none) shared(roi_out, ivoid, ovoid, d) schedule(static)
+#pragma omp parallel for default(none) shared(roi_out, ivoid, ovoid, d, coeffsi) schedule(static)
 #endif
     for(int j=0; j<roi_out->height; j++)
     {
diff --git a/src/iop/tonemap.cc b/src/iop/tonemap.cc
index 4c58d23..53ce612 100644
--- a/src/iop/tonemap.cc
+++ b/src/iop/tonemap.cc
@@ -219,10 +219,10 @@ extern "C"
     else
       module->default_enabled = 0;
 
-    dt_iop_tonemapping_params_t tmp = (dt_iop_tonemapping_params_t)
-    {
-      2.5,0.1
-    };
+    dt_iop_tonemapping_params_t tmp;
+    tmp.contrast = 2.5;
+    tmp.Fsize = 0.1;
+
     memcpy(module->params, &tmp, sizeof(dt_iop_tonemapping_params_t));
     memcpy(module->default_params, &tmp, sizeof(dt_iop_tonemapping_params_t));
   }
diff --git a/src/rawspeed/CMakeLists.txt b/src/rawspeed/CMakeLists.txt
index 16cf282..5316ca9 100644
--- a/src/rawspeed/CMakeLists.txt
+++ b/src/rawspeed/CMakeLists.txt
@@ -48,7 +48,14 @@ set(RAWSPEED_SOURCES
 #
 # build librawspeed
 #
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
+if(CMAKE_COMPILER_IS_GNUCC)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
+endif(CMAKE_COMPILER_IS_GNUCC)
+
+
+if("${CMAKE_C_COMPILER_ID} ${CMAKE_CXX_COMPILER_ID}" MATCHES SunPro)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Kpic")
+endif("${CMAKE_C_COMPILER_ID} ${CMAKE_CXX_COMPILER_ID}" MATCHES SunPro)
 
 add_library(rawspeed STATIC ${RAWSPEED_SOURCES})
 target_link_libraries(rawspeed)
diff --git a/src/rawspeed/RawSpeed/ArwDecoder.h b/src/rawspeed/RawSpeed/ArwDecoder.h
index 6b2d45c..f1d701a 100644
--- a/src/rawspeed/RawSpeed/ArwDecoder.h
+++ b/src/rawspeed/RawSpeed/ArwDecoder.h
@@ -19,7 +19,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef ARW_DECODER_H
+#define ARW_DECODER_H
+
 #include "RawDecoder.h"
 #include "LJpegPlain.h"
 #include "TiffIFD.h"
@@ -46,3 +48,5 @@ protected:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/BitPumpJPEG.h b/src/rawspeed/RawSpeed/BitPumpJPEG.h
index 3e2bc5c..1a31cec 100644
--- a/src/rawspeed/RawSpeed/BitPumpJPEG.h
+++ b/src/rawspeed/RawSpeed/BitPumpJPEG.h
@@ -19,7 +19,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef BIT_PUMP_JPEG_H
+#define BIT_PUMP_JPEG_H
+
 #include "ByteStream.h"
 #include "IOException.h"
 
@@ -117,3 +119,5 @@ private:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/BitPumpMSB.h b/src/rawspeed/RawSpeed/BitPumpMSB.h
index d6fb9c8..8c745cb 100644
--- a/src/rawspeed/RawSpeed/BitPumpMSB.h
+++ b/src/rawspeed/RawSpeed/BitPumpMSB.h
@@ -19,7 +19,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef BIT_PUMP_MSB_H
+#define BIT_PUMP_MSB_H
+
 #include "ByteStream.h"
 
 #define BITS_PER_LONG (8*sizeof(uint32))
@@ -152,3 +154,5 @@ private:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/BitPumpMSB32.h b/src/rawspeed/RawSpeed/BitPumpMSB32.h
index 208bcec..9aab7d8 100644
--- a/src/rawspeed/RawSpeed/BitPumpMSB32.h
+++ b/src/rawspeed/RawSpeed/BitPumpMSB32.h
@@ -19,7 +19,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef BIT_PUMP_MSB32_H
+#define BIT_PUMP_MSB32_H
+
 #include "ByteStream.h"
 
 #ifdef MIN_GET_BITS
@@ -74,3 +76,5 @@ private:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/BitPumpPlain.h b/src/rawspeed/RawSpeed/BitPumpPlain.h
index 4fefdf8..07c063e 100644
--- a/src/rawspeed/RawSpeed/BitPumpPlain.h
+++ b/src/rawspeed/RawSpeed/BitPumpPlain.h
@@ -19,7 +19,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef BIT_PUMP_PLAIN_H
+#define BIT_PUMP_PLAIN_H
+
 #include "ByteStream.h"
 
 namespace RawSpeed {
@@ -54,3 +56,5 @@ private:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/BlackArea.h b/src/rawspeed/RawSpeed/BlackArea.h
index 6f6f6b9..73c92f6 100644
--- a/src/rawspeed/RawSpeed/BlackArea.h
+++ b/src/rawspeed/RawSpeed/BlackArea.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef BLACK_AREA_H
+#define BLACK_AREA_H
+
 /* 
     RawSpeed - RAW file decoder.
 
@@ -33,3 +35,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/ByteStream.h b/src/rawspeed/RawSpeed/ByteStream.h
index 03e743a..d1ca32e 100644
--- a/src/rawspeed/RawSpeed/ByteStream.h
+++ b/src/rawspeed/RawSpeed/ByteStream.h
@@ -19,7 +19,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef BYTE_STREAM_H
+#define BYTE_STREAM_H
+
 #include "IOException.h"
 
 namespace RawSpeed {
@@ -48,3 +50,5 @@ protected:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/ByteStreamSwap.h b/src/rawspeed/RawSpeed/ByteStreamSwap.h
index fef0fc3..1b67eac 100644
--- a/src/rawspeed/RawSpeed/ByteStreamSwap.h
+++ b/src/rawspeed/RawSpeed/ByteStreamSwap.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef BYTE_STREAM_SWAP_H
+#define BYTE_STREAM_SWAP_H
+
 #include "ByteStream.h"
 
 #include "IOException.h"
@@ -17,3 +19,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/Camera.h b/src/rawspeed/RawSpeed/Camera.h
index b8e9573..031ede3 100644
--- a/src/rawspeed/RawSpeed/Camera.h
+++ b/src/rawspeed/RawSpeed/Camera.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef CAMERA_H
+#define CAMERA_H
+
 #include "ColorFilterArray.h"
 #include <libxml/parser.h>
 #include "BlackArea.h"
@@ -58,3 +60,5 @@ protected:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/CameraMetaData.h b/src/rawspeed/RawSpeed/CameraMetaData.h
index fd51c0c..c0d7dd2 100644
--- a/src/rawspeed/RawSpeed/CameraMetaData.h
+++ b/src/rawspeed/RawSpeed/CameraMetaData.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef CAMERA_META_DATA_H
+#define CAMERA_META_DATA_H
+
 #include <libxml/xmlmemory.h>
 #include <libxml/parser.h>
 #include <libxml/xmlschemas.h>
@@ -41,3 +43,5 @@ protected:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/CameraMetadataException.h b/src/rawspeed/RawSpeed/CameraMetadataException.h
index 594e251..55a2814 100644
--- a/src/rawspeed/RawSpeed/CameraMetadataException.h
+++ b/src/rawspeed/RawSpeed/CameraMetadataException.h
@@ -19,7 +19,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef CAMERA_METADATA_EXCEPTION_H
+#define CAMERA_METADATA_EXCEPTION_H
+
 
 namespace RawSpeed {
 
@@ -33,3 +35,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/ColorFilterArray.h b/src/rawspeed/RawSpeed/ColorFilterArray.h
index 46a9c9b..2fc20a9 100644
--- a/src/rawspeed/RawSpeed/ColorFilterArray.h
+++ b/src/rawspeed/RawSpeed/ColorFilterArray.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef COLOR_FILTER_ARRAY_H
+#define COLOR_FILTER_ARRAY_H
+
 #include "RawDecoderException.h"
 /* 
     RawSpeed - RAW file decoder.
@@ -66,3 +68,5 @@ private:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/Common.h b/src/rawspeed/RawSpeed/Common.h
index f269894..d1e7708 100644
--- a/src/rawspeed/RawSpeed/Common.h
+++ b/src/rawspeed/RawSpeed/Common.h
@@ -19,7 +19,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef COMMON_H
+#define COMMON_H
+
 
 #if !defined(__unix__) && !defined(__MINGW32__)
 #include <intrin.h>
@@ -121,3 +123,5 @@ inline uint32 clampbits(int x, uint32 n) { uint32 _y_temp; if( (_y_temp=x>>n) )
 
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/Cr2Decoder.h b/src/rawspeed/RawSpeed/Cr2Decoder.h
index a75a44c..f534b90 100644
--- a/src/rawspeed/RawSpeed/Cr2Decoder.h
+++ b/src/rawspeed/RawSpeed/Cr2Decoder.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef CR2_DECODER_H
+#define CR2_DECODER_H
+
 #include "RawDecoder.h"
 #include "LJpegPlain.h"
 #include "TiffIFD.h"
@@ -55,4 +57,5 @@ public:
   uint32 count;
 };
 
-} // namespace RawSpeed
\ No newline at end of file
+} // namespace RawSpeed
+#endif
diff --git a/src/rawspeed/RawSpeed/DngDecoder.h b/src/rawspeed/RawSpeed/DngDecoder.h
index d5ff649..187deda 100644
--- a/src/rawspeed/RawSpeed/DngDecoder.h
+++ b/src/rawspeed/RawSpeed/DngDecoder.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef DNG_DECODER_H
+#define DNG_DECODER_H
+
 #include "LJpegPlain.h"
 #include "TiffIFD.h"
 #include "DngDecoderSlices.h"
@@ -55,3 +57,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/DngDecoderSlices.h b/src/rawspeed/RawSpeed/DngDecoderSlices.h
index 126c99d..a915e3c 100644
--- a/src/rawspeed/RawSpeed/DngDecoderSlices.h
+++ b/src/rawspeed/RawSpeed/DngDecoderSlices.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef DNG_DECODER_SLICES_H
+#define DNG_DECODER_SLICES_H
+
 #include "RawDecoder.h"
 #include <queue>
 #include "LJpegPlain.h"
@@ -72,3 +74,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/FileIOException.h b/src/rawspeed/RawSpeed/FileIOException.h
index ced5dc0..6b61da1 100644
--- a/src/rawspeed/RawSpeed/FileIOException.h
+++ b/src/rawspeed/RawSpeed/FileIOException.h
@@ -21,7 +21,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef FILE_IOEXCEPTION_H
+#define FILE_IOEXCEPTION_H
+
 
 namespace RawSpeed {
 
@@ -34,3 +36,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/FileMap.h b/src/rawspeed/RawSpeed/FileMap.h
index 93129c7..5c15918 100644
--- a/src/rawspeed/RawSpeed/FileMap.h
+++ b/src/rawspeed/RawSpeed/FileMap.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef FILE_MAP_H
+#define FILE_MAP_H
+
 #include "FileIOException.h"
 /* 
     RawSpeed - RAW file decoder.
@@ -55,3 +57,5 @@ private:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/FileReader.h b/src/rawspeed/RawSpeed/FileReader.h
index d245ff0..7448ec2 100644
--- a/src/rawspeed/RawSpeed/FileReader.h
+++ b/src/rawspeed/RawSpeed/FileReader.h
@@ -19,7 +19,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef FILE_READER_H
+#define FILE_READER_H
+
 #include "FileIOException.h"
 #include "FileMap.h"
 
@@ -39,3 +41,5 @@ private:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/IOException.h b/src/rawspeed/RawSpeed/IOException.h
index 85924dd..94ce381 100644
--- a/src/rawspeed/RawSpeed/IOException.h
+++ b/src/rawspeed/RawSpeed/IOException.h
@@ -21,7 +21,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef IOEXCEPTION_H
+#define IOEXCEPTION_H
+
 
 namespace RawSpeed {
 
@@ -36,3 +38,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/LJpegDecompressor.h b/src/rawspeed/RawSpeed/LJpegDecompressor.h
index c031a49..24978eb 100644
--- a/src/rawspeed/RawSpeed/LJpegDecompressor.h
+++ b/src/rawspeed/RawSpeed/LJpegDecompressor.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef LJPEG_DECOMPRESSOR_H
+#define LJPEG_DECOMPRESSOR_H
+
 #include "RawDecoder.h"
 #include "BitPumpMSB.h"
 /* 
@@ -194,3 +196,5 @@ protected:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/LJpegPlain.h b/src/rawspeed/RawSpeed/LJpegPlain.h
index 395658a..c186240 100644
--- a/src/rawspeed/RawSpeed/LJpegPlain.h
+++ b/src/rawspeed/RawSpeed/LJpegPlain.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef LJPEG_PLAIN_H
+#define LJPEG_PLAIN_H
+
 #include "LJpegDecompressor.h"
 #include "BitPumpMSB.h"
 /* 
@@ -49,3 +51,5 @@ private:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/NefDecoder.h b/src/rawspeed/RawSpeed/NefDecoder.h
index b034cb1..a1f7496 100644
--- a/src/rawspeed/RawSpeed/NefDecoder.h
+++ b/src/rawspeed/RawSpeed/NefDecoder.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef NEF_DECODER_H
+#define NEF_DECODER_H
+
 #include "RawDecoder.h"
 #include "LJpegPlain.h"
 #include "TiffIFD.h"
@@ -55,4 +57,5 @@ public:
   uint32 count;
 };
 
-} // namespace RawSpeed
\ No newline at end of file
+} // namespace RawSpeed
+#endif
diff --git a/src/rawspeed/RawSpeed/NikonDecompressor.h b/src/rawspeed/RawSpeed/NikonDecompressor.h
index adbef86..fdec584 100644
--- a/src/rawspeed/RawSpeed/NikonDecompressor.h
+++ b/src/rawspeed/RawSpeed/NikonDecompressor.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef NIKON_DECOMPRESSOR_H
+#define NIKON_DECOMPRESSOR_H
+
 #include "LJpegDecompressor.h"
 #include "BitPumpMSB.h"
 /* 
@@ -55,3 +57,5 @@ static const uchar8 nikon_tree[][32] = {
   7,6,8,5,9,4,10,3,11,12,2,0,1,13,14 } };
   
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/OrfDecoder.h b/src/rawspeed/RawSpeed/OrfDecoder.h
index d7d8f52..ff7c535 100644
--- a/src/rawspeed/RawSpeed/OrfDecoder.h
+++ b/src/rawspeed/RawSpeed/OrfDecoder.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef ORF_DECODER_H
+#define ORF_DECODER_H
+
 #include "RawDecoder.h"
 #include "LJpegPlain.h"
 #include "TiffIFD.h"
@@ -43,3 +45,5 @@ private:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/PefDecoder.h b/src/rawspeed/RawSpeed/PefDecoder.h
index 4227ea3..59ca83b 100644
--- a/src/rawspeed/RawSpeed/PefDecoder.h
+++ b/src/rawspeed/RawSpeed/PefDecoder.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef PEF_DECODER_H
+#define PEF_DECODER_H
+
 #include "RawDecoder.h"
 #include "TiffIFD.h"
 #include "PentaxDecompressor.h"
@@ -39,3 +41,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/PentaxDecompressor.h b/src/rawspeed/RawSpeed/PentaxDecompressor.h
index 2090737..69adc69 100644
--- a/src/rawspeed/RawSpeed/PentaxDecompressor.h
+++ b/src/rawspeed/RawSpeed/PentaxDecompressor.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef PENTAX_DECOMPRESSOR_H
+#define PENTAX_DECOMPRESSOR_H
+
 #include "LJpegDecompressor.h"
 #include "BitPumpMSB.h"
 #include "TiffIFD.h"
@@ -39,3 +41,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/RawDecoder.h b/src/rawspeed/RawSpeed/RawDecoder.h
index e0302f3..7d44b4d 100644
--- a/src/rawspeed/RawSpeed/RawDecoder.h
+++ b/src/rawspeed/RawSpeed/RawDecoder.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef RAW_DECODER_H
+#define RAW_DECODER_H
+
 #include "RawDecoderException.h"
 #include "FileMap.h"
 #include "BitPumpJPEG.h" // Includes bytestream
@@ -147,3 +149,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/RawDecoderException.h b/src/rawspeed/RawSpeed/RawDecoderException.h
index edd9c06..d737c96 100644
--- a/src/rawspeed/RawSpeed/RawDecoderException.h
+++ b/src/rawspeed/RawSpeed/RawDecoderException.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef RAW_DECODER_EXCEPTION_H
+#define RAW_DECODER_EXCEPTION_H
+
 /* 
     RawSpeed - RAW file decoder.
 
@@ -34,3 +36,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/RawImage.h b/src/rawspeed/RawSpeed/RawImage.h
index a0faa32..5a8cf28 100644
--- a/src/rawspeed/RawSpeed/RawImage.h
+++ b/src/rawspeed/RawSpeed/RawImage.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef RAW_IMAGE_H
+#define RAW_IMAGE_H
+
 #include "ColorFilterArray.h"
 #include "BlackArea.h"
 
@@ -105,3 +107,5 @@ inline RawImage RawImage::create(iPoint2D dim, uint32 bytesPerPixel, uint32 comp
 { return new RawImageData(dim, bytesPerPixel, componentsPerPixel); }
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/Rw2Decoder.h b/src/rawspeed/RawSpeed/Rw2Decoder.h
index a6cea5b..435b236 100644
--- a/src/rawspeed/RawSpeed/Rw2Decoder.h
+++ b/src/rawspeed/RawSpeed/Rw2Decoder.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef RW2_DECODER_H
+#define RW2_DECODER_H
+
 #include "RawDecoder.h"
 #include "TiffIFD.h"
 #include "BitPumpPlain.h"
@@ -60,3 +62,5 @@ private:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/TiffEntry.h b/src/rawspeed/RawSpeed/TiffEntry.h
index 69ba86a..58fae2f 100644
--- a/src/rawspeed/RawSpeed/TiffEntry.h
+++ b/src/rawspeed/RawSpeed/TiffEntry.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef TIFF_ENTRY_H
+#define TIFF_ENTRY_H
+
 #include "TiffParserException.h"
 #include "FileMap.h"
 
@@ -92,3 +94,5 @@ protected:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/TiffEntryBE.h b/src/rawspeed/RawSpeed/TiffEntryBE.h
index 7453bc5..fc1acc1 100644
--- a/src/rawspeed/RawSpeed/TiffEntryBE.h
+++ b/src/rawspeed/RawSpeed/TiffEntryBE.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef TIFF_ENTRY_BE_H
+#define TIFF_ENTRY_BE_H
+
 #include "TiffEntry.h"
 
 /* 
@@ -41,3 +43,5 @@ private:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/TiffIFD.h b/src/rawspeed/RawSpeed/TiffIFD.h
index 0b73e4c..4238581 100644
--- a/src/rawspeed/RawSpeed/TiffIFD.h
+++ b/src/rawspeed/RawSpeed/TiffIFD.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef TIFF_IFD_H
+#define TIFF_IFD_H
+
 #include "FileMap.h"
 #include "TiffEntry.h"
 #include "TiffParserException.h"
@@ -68,3 +70,5 @@ inline Endianness getTiffEndianness(const ushort16* tifftag) {
 }
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/TiffIFDBE.h b/src/rawspeed/RawSpeed/TiffIFDBE.h
index 9579187..a5488c8 100644
--- a/src/rawspeed/RawSpeed/TiffIFDBE.h
+++ b/src/rawspeed/RawSpeed/TiffIFDBE.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef TIFF_IFDBE_H
+#define TIFF_IFDBE_H
+
 #include "TiffIFD.h"
 /* 
     RawSpeed - RAW file decoder.
@@ -34,3 +36,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/TiffParser.h b/src/rawspeed/RawSpeed/TiffParser.h
index 890446c..f20ce9c 100644
--- a/src/rawspeed/RawSpeed/TiffParser.h
+++ b/src/rawspeed/RawSpeed/TiffParser.h
@@ -20,7 +20,9 @@
     http://www.klauspost.com
 */
 
-#pragma once
+#ifndef TIFF_PARSER_H
+#define TIFF_PARSER_H
+
 #include "FileMap.h"
 #include "TiffIFD.h"
 #include "TiffIFDBE.h"
@@ -54,3 +56,5 @@ protected:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/TiffParserException.h b/src/rawspeed/RawSpeed/TiffParserException.h
index 925eab7..c3f3889 100644
--- a/src/rawspeed/RawSpeed/TiffParserException.h
+++ b/src/rawspeed/RawSpeed/TiffParserException.h
@@ -19,7 +19,9 @@
 
     http://www.klauspost.com
 */
-#pragma once
+#ifndef TIFF_PARSER_EXCEPTION_H
+#define TIFF_PARSER_EXCEPTION_H
+
 
 namespace RawSpeed {
 
@@ -32,3 +34,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/TiffParserHeaderless.h b/src/rawspeed/RawSpeed/TiffParserHeaderless.h
index 1d31c62..867ca7a 100644
--- a/src/rawspeed/RawSpeed/TiffParserHeaderless.h
+++ b/src/rawspeed/RawSpeed/TiffParserHeaderless.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef TIFF_PARSER_HEADERLESS_H
+#define TIFF_PARSER_HEADERLESS_H
+
 #include "TiffParser.h"
 /* 
     RawSpeed - RAW file decoder.
@@ -35,3 +37,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/src/rawspeed/RawSpeed/TiffParserOlympus.h b/src/rawspeed/RawSpeed/TiffParserOlympus.h
index f7ba283..ae95698 100644
--- a/src/rawspeed/RawSpeed/TiffParserOlympus.h
+++ b/src/rawspeed/RawSpeed/TiffParserOlympus.h
@@ -1,4 +1,6 @@
-#pragma once
+#ifndef TIFF_PARSER_OLYMPUS_H
+#define TIFF_PARSER_OLYMPUS_H
+
 #include "TiffParser.h"
 /* 
     RawSpeed - RAW file decoder.
@@ -34,3 +36,5 @@ public:
 };
 
 } // namespace RawSpeed
+
+#endif
diff --git a/tools/create_metadata.sh b/tools/create_metadata.sh
index a67138b..96a0b90 100755
--- a/tools/create_metadata.sh
+++ b/tools/create_metadata.sh
@@ -5,7 +5,7 @@ outputdir=$2
 outputheader=metadata_gen.h
 outputbody=metadata_gen.c
 
-headerdefine=__$(echo -n $outputheader | tr '[:lower:].' '[:upper:]_')__
+headerdefine=__$(printf $outputheader | tr '[:lower:].' '[:upper:]_')__
 # header of the .h file
 cat > $outputdir/$outputheader << EOF
 /** generated file, do not edit! */
@@ -31,12 +31,12 @@ EOF
 # iterate over the input
 first=0
 for line in $(cat $inputfile | grep -v "^#"); do
-    enum=DT_METADATA_$(echo -n $line | tr '[:lower:].' '[:upper:]_')
-    length=$(echo -n $line | wc -c)
+    enum=DT_METADATA_$(printf $line | tr '[:lower:].' '[:upper:]_')
+    length=$(printf $line | wc -c)
     if [ "$first" -ne 0 ]; then
         echo "," >> $outputdir/$outputheader
     fi
-    echo -n "    $enum" >> $outputdir/$outputheader
+    printf "    $enum" >> $outputdir/$outputheader
     first=1
 
     cat >> $outputdir/$outputbody << EOF
-- 
1.5.6.5

